src/wiktextract/extractor/es/models.py

```diff
@@ -0,0 +1,113 @@
+from typing import Optional
+import json
+
+import logging
+
+from pydantic import BaseModel, Field, model_validator
+from pydantic.json_schema import GenerateJsonSchema
+
+from wiktextract.wxr_context import WiktextractContext
+
+
+class PydanticLogger:
+    wxr: Optional[WiktextractContext] = None
+
+    @classmethod
+    def debug(
+        cls, msg: str, trace: Optional[str] = None, sortid: str = "XYZunsorted"
+    ):
+        if cls.wxr:
+            cls.wxr.wtp.debug(msg, trace=trace, sortid=sortid)
+        else:
+            logging.debug(msg)
+
+
+class BaseModelWrap(BaseModel):
+    class Config:
+        extra = "ignore"
+        validate_assignment = True
+
+    def update(self, data: dict):
+        update = self.dict(exclude_defaults=True, exclude_none=True)
+        update.update(data)
+        for k, v in (
+            self.validate(update)
+            .dict(exclude_defaults=True, exclude_none=True)
+            .items()
+        ):
+            setattr(self, k, v)
+        return self
+
+
+class LoggingExtraFieldsModel(BaseModelWrap):
+    @model_validator(mode="before")
+    def log_extra_fields(cls, values):
+        all_allowed_field_names = {key for key in cls.__fields__.keys()}
+        extra_fields = {
+            name: str(value)
+            for name, value in values.items()
+            if name not in all_allowed_field_names
+        }
+        if extra_fields:
+            class_full_name = cls.__name__
+            PydanticLogger.debug(
+                msg=f"Pydantic - Got extra fields in {class_full_name}: {extra_fields}",
+                sortid="wiktextract/extractor/es/pydantic/extra_fields/33",
+            )
+        return values
```

Comment:
Since extra fields are ignored by default, we might be interested in logging when we _accidentally_ add undefined fields. This might be useful, for example, when dynamically adding key value pairs from a wiki template.